core/src/main/java/org/apache/spark/shuffle/sort/ShuffleExternalSorter.java

```diff
@@ -206,14 +211,21 @@ private void writeSortedFile(boolean isLastFile) {
       long recordReadPosition = recordOffsetInPage + uaoSize; // skip over record length
       while (dataRemaining > 0) {
         final int toTransfer = Math.min(diskWriteBufferSize, dataRemaining);
-        Platform.copyMemory(
-          recordPage, recordReadPosition, writeBuffer, Platform.BYTE_ARRAY_OFFSET, toTransfer);
-        writer.write(writeBuffer, 0, toTransfer);
+        if (bufferOffset > 0 && bufferOffset + toTransfer > DISK_WRITE_BUFFER_SIZE) {
```

Comment:
Not a bad idea, but codes here may not work as you expect. If we got a record with size `X` < `diskWriteBufferSize `(same as `DISK_WRITE_BUFFER_SIZE `), then we will only call `writer.write()` once. And if we got a record with size `Y` >= `diskWriteBufferSize `, then we will call `writer.write()` for  (`Y` + `diskWriteBufferSize ` - 1)  / `diskWriteBufferSize`  times. And this do not change with the new code.

Reply:
Thanks @Ngone51 
If we got `n` records with size n*X < diskWriteBufferSize(same as DISK_WRITE_BUFFER_SIZE), then we will only call writer.write() once.

Reply:
Oh, I see. If so, I'm afraid you may have to change ` writer.recordWritten()`'s behaviour, which just count records one bye one right now.

Reply:
Actually, it doesn't need to change.
The result  `numRecordsWritten` has no effect,  tt's only written in `diskWriteBuffer` before, but now it's written to `writeBuffer`.
The bytes which has Written will be updated in `commitAndGet`

Reply:
Yeah, I agree there' s no difference as for final result. But `writeMetrics` in `DiskBlockObjectWriter` would be incorrect during the process. Not only `numRecordsWritten`, but also `_bytesWritten`(this could only be correctly counted when `writer.write()` is called. You can see `recordWritten#updateBytesWritten` for detail).

Reply:
The `numRecordsWritten`  in `DiskBlockObjectWriter`  is still  correct during the process after this PR

Reply:
> The numRecordsWritten in DiskBlockObjectWriter is still correct during the process after this PR

The number is correct, but it is not consistent with what real happen compare to current behaviour.  But as you said, we will get correct result at the end. So, it may not be a big deal.

Reply:
Yes,you are right.